src/services/oss/backend.rs

```diff
@@ -0,0 +1,655 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::io::Result;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use http::header;
+use http::Request;
+use http::Response;
+use http::StatusCode;
+use http::Uri;
+use reqsign::services::aliyun::oss::Signer;
+
+use super::dir_stream::DirStream;
+use super::error::parse_error;
+use crate::accessor::AccessorCapability;
+use crate::error::new_other_backend_error;
+use crate::error::new_other_object_error;
+use crate::http_util::new_request_build_error;
+use crate::http_util::new_request_send_error;
+use crate::http_util::new_request_sign_error;
+use crate::http_util::new_response_consume_error;
+use crate::http_util::parse_content_length;
+use crate::http_util::parse_error_response;
+use crate::http_util::parse_etag;
+use crate::http_util::parse_last_modified;
+use crate::http_util::percent_encode_path;
+use crate::http_util::AsyncBody;
+use crate::http_util::HttpClient;
+use crate::http_util::IncomingAsyncBody;
+use crate::object::ObjectPageStreamer;
+use crate::ops::BytesRange;
+use crate::ops::OpCreate;
+use crate::ops::OpDelete;
+use crate::ops::OpList;
+use crate::ops::OpRead;
+use crate::ops::OpStat;
+use crate::ops::OpWrite;
+use crate::ops::Operation;
+use crate::path::build_abs_path;
+use crate::path::normalize_root;
+use crate::Accessor;
+use crate::AccessorMetadata;
+use crate::BytesReader;
+use crate::ObjectMetadata;
+use crate::ObjectMode;
+use crate::ObjectStreamer;
+use crate::Scheme;
+
+type AsyncReq = Request<AsyncBody>;
+type AsyncResp = Response<IncomingAsyncBody>;
+
+const OSS_DEFAULT_ENDPOINT_HOST: &str = "oss-accelerate.aliyuncs.com";
+
+/// Builder for Aliyun Object Storage Service
+#[derive(Default, Clone)]
+pub struct Builder {
+    root: Option<String>,
+
+    region: Option<String>,
```

Comment:
region should be included in endpoint.

Reply:
OSS backend only needs one of the `endpoint` or `region` fields.

Users can set `region` only, like `oss-cn-beijing` or `oss-cn-hangzhou`, then the OSS backend will automatically access `https://<bucket>.oss-cn-beijing.aliyuncs.com` or `https://<bucket>.oss-cn-hangzhou.aliyuncs.com`.

Users can also set `endpoint`, and then `region` will be ignored and the backend will access `https://<bucket>.<endpoint>`.

I prefer keeping `region` because this allows users to type less:

- `OPENDAL_OSS_REGION=oss-cn-beijing`
- `OPENDAL_OSS_ENDPOINT=https://oss-cn-beijing.aliyuncs.com`

these are equal, but the first one is more convenient.

Reply:
Negative. OSS users don't know about regions. (even in their own SDK)

Reply:
Ok, I'll remove this to stay aligned with their SDKs.